core/cog-gamepad-manette.c

```diff
@@ -0,0 +1,343 @@
+{
+    uint16_t                button;
+    enum wpe_gamepad_button std_button;
+
+    if (!manette_event_get_button(event, &button))
+        return;
+
+    if (to_standard_gamepad_button(button, &std_button))
+        wpe_gamepad_dispatch_button_changed(gamepad->wk_gamepad, std_button, pressed);
+}
+
+static void
+on_button_press(ManetteDevice *device, ManetteEvent *event, Gamepad *gamepad)
+{
+    dispatch_button_event(gamepad, event, true);
+}
+
+static void
+on_button_release(ManetteDevice *device, ManetteEvent *event, Gamepad *gamepad)
+{
+    dispatch_button_event(gamepad, event, false);
+}
+
+static void
+device_disconnect_cb(ManetteDevice *device, Gamepad *gamepad)
+{
+    if (!gamepad)
+        return;
+
+    g_signal_handlers_disconnect_by_data(gamepad->device, gamepad);
+}
+
+static gpointer
+gamepad_create(struct wpe_gamepad *wk_gamepad, struct wpe_gamepad_provider *wk_provider, unsigned gamepad_id)
+{
+    Gamepad       *gamepad;
+    ManetteDevice *device;
+    Provider      *provider;
+
+    provider = wpe_gamepad_provider_get_backend(wk_provider);
+    g_assert(provider);
+    device = g_hash_table_lookup(provider->devices, GUINT_TO_POINTER(gamepad_id));
+    g_assert(MANETTE_IS_DEVICE(device));
+
+    gamepad = g_new0(Gamepad, 1);
+
+    gamepad->wk_gamepad = wk_gamepad;
+    gamepad->device = g_object_ref(device);
+
+    g_signal_connect(device, "button-press-event", G_CALLBACK(on_button_press), gamepad);
+    g_signal_connect(device, "button-release-event", G_CALLBACK(on_button_release), gamepad);
+    g_signal_connect(device, "absolute-axis-event", G_CALLBACK(on_absolute_axis), gamepad);
+    g_signal_connect(device, "disconnected", G_CALLBACK(device_disconnect_cb), gamepad);
+
+    g_debug("gamepad: %s - created", manette_device_get_name(device));
+
+    return gamepad;
+}
+
+static void
+gamepad_destroy(gpointer data)
+{
+    Gamepad *gamepad = data;
+
+    if (!gamepad)
+        return;
+
+    g_clear_object(&gamepad->device);
+    g_free(gamepad);
+}
+
+static const char *
+gamepad_get_id(gpointer data)
+{
+    Gamepad *gamepad = data;
+
+    if (!(gamepad && gamepad->device))
+        return "Unknown";
+
+    return manette_device_get_name(gamepad->device);
+}
+
+static struct wpe_gamepad_provider_interface provider_interface = {
+    .create = provider_create,
+    .destroy = provider_destroy,
+    .start = provider_start,
+    .stop = provider_stop,
+};
+
+static struct wpe_gamepad_interface device_interface = {
+    .create = gamepad_create,
+    .destroy = gamepad_destroy,
+    .get_id = gamepad_get_id,
+};
+
+/**
+ * cog_register_gamepad_backend:
+ * @provider_get_view_for_gamepad: A function that * returns the view backend
+ *    associated with the gamepad.
+ *
+ * Installs a gamepad backend, which is in charge of the physical device
+ * events.
+ */
+void
+cog_register_gamepad_backend(GamepadProviderGetViewBackend *provider_get_view_for_gamepad)
```

Comment:
assuming that users will have only one gamepad implementation it will choose at configuration/compilation time, thus only one `cog_register_gamepad_backend` will be resolved by the linker. Right now there are two implementations: noop and manette.

Reply:
I don't see why we wouldn't allow chosing an implementation at runtime, the same way we allow choosing Wayland/X11/etc. at runtime. It's not like it would be much more work anyway 🤔

Reply:
I would leave this change for a next iteration, since it's local to cog.

Reply:
That would be fine 👌🏼